[Zhu Xihua] iP

XihuaZ opened the pull request:
### Bot

> 'If you cannot do great things, do small things in a great way.' – Napoleon Hill

- text-based
- easy to learn
- ~~SLOW~~ FAST running speed

Make sure you have the following available:

1. Java

You can do:

- [ ] Add upcoming tasks
- [ ] Mark and unmark them
- [ ] Search for tasks
- [ ] Look at all tasks

What are you waiting for? Download [now](https://github.com/XihuaZ/ip/releases) 😉

This is the `Launcher` class you will need to use:

```
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}

```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

XihuaZ committed:
Level 0. Rename, Greet, Exit

XihuaZ committed:
Level 1. Echo

XihuaZ committed:
Level 2. Add, List

XihuaZ committed:
Level 3. Mark as Done

XihuaZ committed:
Level 4. ToDos, Events, Deadlines

XihuaZ committed:
A-TextUiTesting: Automated Text UI Testing

XihuaZ committed:
Level 5. Handle Errors

XihuaZ committed:
Level 6. Delete

XihuaZ committed:
Cleanup redundant code

XihuaZ committed:
Level 7. Save

XihuaZ committed:
Implemented conversion to LocalDateTime

XihuaZ committed:
Level 8. Dates and Times

XihuaZ committed:
Refactor the code

XihuaZ committed:
Organize into Packages

XihuaZ committed:
Modify build.gradle file

XihuaZ committed:
Add Junit tests and modify Ui.java

XihuaZ committed:
Debug JUnit tests and update build.gradle

ZD292 commented on src/main/java/bot/Ui.java:
```diff
@@ -0,0 +1,51 @@
+package bot;
```
Looks good to me 😄

itssisi commented on src/main/java/bot/storage/Storage.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.Iterator;
+
+public class Storage {
+    private static final String FILE_NAME = "task.txt";
+    private static final String FILE_RELATIVE_PATH = "data" + File.separator;
+    private static final String WORKING_DIRECTORY = System.getProperty("user.dir");
+    private static final String ABSOLUTE_FILE_PATH = Storage.WORKING_DIRECTORY + File.separator +
+            Storage.FILE_RELATIVE_PATH;
```
Good SE practice in constant naming convention.

itssisi commented on src/main/java/bot/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,44 @@
+                    "Got it. I've added this bot.task:\n" +
+                    this.deadline + "\n" +
+                    "Now you have " + this.taskList.length() + " tasks in the list.\n" +
+                    Command.SPACER;
```
Good implementation of separation of lines when the code is too long.

itssisi commented on src/main/java/bot/command/MarkCommand.java:
```diff
@@ -0,0 +1,34 @@
+
+
+public class MarkCommand extends Command {
+    private final Task task;
+    private final TaskList taskList;
```
Is final necessary?

itssisi commented on src/main/java/bot/ControlFlow.java:
```diff
@@ -0,0 +1,120 @@
+        this.taskList = taskList;
+    }
+
+
```
Should Javadoc comment be written here?

itssisi commented on src/main/java/bot/storage/Storage.java:
```diff
@@ -0,0 +1,61 @@
+        } catch (IOException exception) {
+            throw new FileErrorBotException("File cannot be saved...");
+        }
+    }
```
Should Javadoc explain the method here?

itssisi commented on src/main/java/bot/command/EventCommand.java:
```diff
@@ -0,0 +1,46 @@
+                    "Now you have " + this.taskList.length() + " tasks in the list.\n" +
+                    Command.SPACER;
+        }
+    }
```
Good implementation of Egyptian braces rule.

itssisi commented:
I think generally the SE coding practices are well-implemented. However, there are missing Javadocs comments and headers in all the files. I think it will be good to add the comments to explain the methods and classes.

XihuaZ committed:
Add JavaDoc comments to the code

XihuaZ committed:
Modify code to comply with coding standard

XihuaZ committed:
Add option to find tasks containing keyword/pattern

XihuaZ committed:
Merge branch 'branch-A-CodingStandard' into branch-Level-9

XihuaZ committed:
Resolve merge conflict by incorporating A-JavaDoc suggestion

XihuaZ committed:
Resolve merge conflict by incorporating branch-Level-9 suggestions

nicholastngyk commented on src/main/java/bot/ControlFlow.java:
```diff
@@ -0,0 +1,133 @@
+        if (str.isBlank()) {
+            throw new IncompleteBotException("OOPS!!! There is no bot.task specified.");
+        }
+        String[] strSplit = InputParser.getSplitAtSpace(str);
+        String taskName = InputParser.getLeftOfSplit(strSplit);
+        String remainder = "";
+        if (strSplit.length > 1) {
+            remainder = InputParser.getRightOfSplit(strSplit);
+        }
+        String[] taskSplit;
+        String taskDetail;
+        String timeline;
+        String[] timelineArr;
+        String timeFrom;
+        String timeTo;
```
Since the method is so long, you might want to utilise blank spaces between rows to structure the segments of code while making the code easier to read.

nicholastngyk commented on src/main/java/bot/command/Command.java:
```diff
@@ -0,0 +1,19 @@
+
+public abstract class Command {
+
+    public static final String SPACER = "____________________________________________________________";
```
Since spacer is only used in commands, you might consider making it protected to limit the use of spacer to within subclasses.

nicholastngyk commented on src/main/java/bot/exception/IncompleteBotException.java:
```diff
@@ -0,0 +1,13 @@
+package bot.exception;
+
+public class IncompleteBotException extends BotException {
```
Good use of different exceptions to categorise the various types of exceptions you might encounter.

nicholastngyk commented on src/main/java/bot/parsers/InputParser.java:
```diff
@@ -0,0 +1,69 @@
+package bot.parsers;
+
+public class InputParser {
+    private static final String SPLIT_AT_SPACE = " ";
+    private static final String SPLIT_AT_BY = " /by ";
+    private static final String SPLIT_AT_FROM = " /from ";
+    private static final String SPLIT_AT_TO = " /to ";
```
Just a small consideration, but by including a space at by, from and to, would it cause an exception for inputs. such as "deadline /by22/08/13 ..." ? Could look at splitting with "/by" then trimming.

nicholastngyk commented on src/main/java/bot/task/Task.java:
```diff
@@ -0,0 +1,82 @@
+    private static final String INCOMPLETE = "[ ] ";
+    protected static final String UNIQUE_FILE_SEPARATOR = " &##& ";
+    private final String name;
+    private boolean isDone;
```
Good naming convention, made it easy to follow

nicholastngyk reviewed (commented):
Overall, a good product with no major errors, just some minor attributes that can (but not needed) to be changed. Good job!

XihuaZ committed:
Include images in resources for GUI

XihuaZ committed:
Include GUI for app

XihuaZ committed:
Fix file directory bug

XihuaZ committed:
MainWindow class: include app termination function

User can continue interacting with the chatbot after "bye" was entered.

This deviates from the intended functionality, where the app is supposed
to terminate when "bye" is inputted.

When a "bye" message is received, a Terminate Command is outputted to
the handleUserInput() method in MainWindow, which will trigger
Platform.exit(). This will close the entire app itself.

It is done this way as execute() is supposed to return a String. Hence,
this function is left to MainWindow class to detect for the Terminate
Command instead.

XihuaZ committed:
Include assert statements

XihuaZ committed:
Improve overall code quality

There is a method that is too lengthy.

The content within the method is moved into the respective command.

This allows for the method's readability to improve, allowing for better
abstraction. Hence, allowing for better conformant with SLAP.

XihuaZ committed:
Merge branch 'A-CodeQuality'

XihuaZ committed:
Include user guide and a picture of the UI

XihuaZ committed:
Update README.md to include address to latest jar file

XihuaZ committed:
Fix grammatical errors to conform with coding standard

XihuaZ committed:
Improve User guide to be more detailed.

XihuaZ committed:
Include command for finding tasks not overdue

XihuaZ committed:
Modify into finding tasks not done within datetime

XihuaZ committed:
Modify javadocs

XihuaZ committed:
Update README.md to include new feature

XihuaZ committed:
Update Ui.png to showcase the Bot chatbot better

XihuaZ committed:
Improve the UG sentence structure

XihuaZ committed:
Merge pull request #2 from XihuaZ/branch-PR-utilise

Improve the UG sentence structure

XihuaZ committed:
Fix error with output sentence

XihuaZ committed:
Merge remote-tracking branch 'refs/remotes/origin/master'
